server/athenian/api/controllers/features/github/pull_request_metrics.py

```diff
@@ -464,6 +463,24 @@ def _analyze(self, facts: pd.DataFrame, min_time: datetime, max_time: datetime,
         return result
 
 
+@register_metric(PullRequestMetricID.PR_REVIEWED)
+class ReviewedCalculator(SumMetricCalculator[int]):
+    """Number of reviewed PRs."""
+
+    dtype = int
+
+    def _analyze(self, facts: np.ndarray, min_time: datetime, max_time: datetime,
+                 **kwargs) -> np.ndarray:
+        dtype = facts["created"].dtype
+        min_time = np.array(min_time, dtype=dtype)
+        max_time = np.array(max_time, dtype=dtype)
+        reviews = np.array(facts["reviews"])
+        result = np.vectorize(
```

Comment:
I wasn't able to find any smart ways for using vectorization here 😞, maybe you @vmarkovtsev can. I thought about different options such as using `df.apply`, but that's basically the same as `np.vectorize` and I didn't know how to proceed with your approach of using `np.cumsum`. I thought about also to build a boolean matrix where the number of cols is given by the max number of reviews in a PR, but that would have required iterating on each row anyways since `np.vstack` do not accept arrays with different sizes by automatically expanding them and filling with empty value.

Reply:
I did a similar trick in the past, let me sketch it.

Reply:
Here it is:

```
@register_metric(PullRequestMetricID.PR_REVIEWED)
class ReviewedCalculator(SumMetricCalculator[int]):
    """Number of reviewed PRs."""

    def _analyze(self, facts: np.ndarray, min_time: datetime, max_time: datetime,
                 **kwargs) -> np.ndarray:
        """Calculate the actual state update."""
        dtype = facts["created"].dtype
        min_time = np.array(min_time, dtype=dtype)
        max_time = np.array(max_time, dtype=dtype)
        review_timestamps = np.concatenate(facts["reviews"])
        reviews_in_range = (min_time <= review_timestamps) & (review_timestamps < max_time)
        # we cannot sum `reviews_in_range` because there can be several reviews for the same PR
        review_offsets = np.zeros(len(facts) + 1, dtype=int)
        review_offsets[1:] = np.cumsum(facts["reviews"].apply(len).values)
        # np.searchsorted aliases several reviews of the same PR to the right border of a `review_offsets` interval
        # np.unique collapses duplicate indexes
        reviewed_indexes = np.unique(np.searchsorted(
            review_offsets, np.where(reviews_in_range)[0], side="right") - 1)
        result = np.full(len(facts), False)
        result[reviewed_indexes] = True
        return result
```

Reply:
Awesome! This is twice faster!